File: push/src/main/java/com/xiaomi/slim/BlobReader.java
package com.xiaomi.slim;

import com.xiaomi.channel.commonutils.logger.MyLog;
import com.xiaomi.channel.commonutils.misc.DebugUtils;
import com.xiaomi.push.protobuf.ChannelMessage;
import com.xiaomi.push.service.PushClientsManager;
import com.xiaomi.push.service.PushSlimHandshakePlan;
import com.xiaomi.push.service.PushSlimPayloadPlan;
import io.github.magisk317.mipush.service.runtime.PushSlimStreamRuntime;
import com.xiaomi.push.service.RC4Cryption;
import java.io.BufferedInputStream;
import java.io.EOFException;
import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.util.zip.Adler32;

/* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/slim/BlobReader.class */
class BlobReader {
    private static final int MAX_BLOB_THRSHOLD = 4096;
    private static final int MIN_BLOB_SIZE = 2048;
    private SlimConnection mConnection;
    private volatile boolean mDone;
    private InputStream mInputStream;
    private byte[] mKey;
    private ByteBuffer mBuffer = ByteBuffer.allocate(MIN_BLOB_SIZE);
    private ByteBuffer mCRCBuf = ByteBuffer.allocate(4);
    private Adler32 mChecksumTool = new Adler32();
    private PacketParser mPacketParser = new PacketParser();

    BlobReader(InputStream inputStream, SlimConnection slimConnection) {
        this.mInputStream = new BufferedInputStream(inputStream);
        this.mConnection = slimConnection;
    }

    private void loop() throws IOException {
        this.mDone = false;
        boolean z = false;
        Blob blob = read();
        if (Blob.CMD_CONN.equals(blob.getCmd())) {
            ChannelMessage.XMMsgConnResp from = ChannelMessage.XMMsgConnResp.parseFrom(blob.getPayload());
            PushSlimHandshakePlan handshakePlan = PushSlimStreamRuntime.planHandshake(from.hasChallenge(), from.hasPsc());
            z = handshakePlan.getValid();
            if (handshakePlan.getValid()) {
                this.mConnection.setChallenge(from.getChallenge());
            }
            if (handshakePlan.getShouldEmitConfigBlob()) {
                ChannelMessage.PushServiceConfigMsg psc = from.getPsc();
                Blob blob2 = new Blob();
                blob2.setCmd(Blob.CMD_SYNC, Blob.SUBCMD_CONF);
                blob2.setPayload(psc.toByteArray(), null);
                this.mConnection.notifyDataArrived(blob2);
            }
            MyLog.w("[Slim] CONN: host = " + from.getHost());
        }
        if (!z) {
            MyLog.w("[Slim] Invalid CONN");
            throw new IOException("Invalid Connection");
        }
        this.mKey = this.mConnection.getKey();
        while (!this.mDone) {
            Blob blob3 = read();
            this.mConnection.setReadAlive();
            PushSlimPayloadPlan payloadPlan = PushSlimStreamRuntime.planPayloadDispatch(blob3.getPayloadType(), blob3.getCmd(), blob3.getChannelId(), blob3.getSubcmd());
            switch (payloadPlan.getAction()) {
                case DeliverBlob:
                    this.mConnection.notifyDataArrived(blob3);
                    break;
                case ParseSecurePacket:
                    try {
                        this.mConnection.notifyDataArrived(this.mPacketParser.parse(blob3.getDecryptedPayload(PushClientsManager.getInstance().getClientLoginInfoByChidAndUserId(Integer.valueOf(blob3.getChannelId()).toString(), blob3.getFullUserName()).security), this.mConnection));
                    } catch (Exception e) {
                        MyLog.w("[Slim] Parse packet from Blob chid=" + blob3.getChannelId() + "; Id=" + blob3.getPacketID() + " failure:" + e.getMessage());
                    }
                    break;
                case ParsePacket:
                    try {
                        this.mConnection.notifyDataArrived(this.mPacketParser.parse(blob3.getPayload(), this.mConnection));
                    } catch (Exception e2) {
                        MyLog.w("[Slim] Parse packet from Blob chid=" + blob3.getChannelId() + "; Id=" + blob3.getPacketID() + " failure:" + e2.getMessage());
                    }
                    break;
                default:
                    if (payloadPlan.getShouldLogUnknownType()) {
                        MyLog.w("[Slim] unknow blob type " + ((int) blob3.getPayloadType()));
                    }
                    break;
            }
        }
    }

    private void read(ByteBuffer byteBuffer, int i) throws IOException {
        int i2;
        int i3;
        int iPosition = byteBuffer.position();
        do {
            int i4 = this.mInputStream.read(byteBuffer.array(), iPosition, i);
            if (i4 == -1) {
                throw new EOFException();
            }
            i2 = i - i4;
            i3 = iPosition + i4;
            iPosition = i3;
            i = i2;
        } while (i2 > 0);
        byteBuffer.position(i3);
    }

    private ByteBuffer readOnePacket() throws IOException {
        this.mBuffer.clear();
        read(this.mBuffer, 8);
        short s = this.mBuffer.getShort(0);
        short s2 = this.mBuffer.getShort(2);
        if (s != -15618 || s2 != 5) {
            throw new IOException("Malformed Input");
        }
        int i = this.mBuffer.getInt(4);
        int iPosition = this.mBuffer.position();
        if (i > 32768) {
            throw new IOException("Blob size too large");
        }
        if (i + 4 > this.mBuffer.remaining()) {
            ByteBuffer byteBufferAllocate = ByteBuffer.allocate(i + MIN_BLOB_SIZE);
            byteBufferAllocate.put(this.mBuffer.array(), 0, this.mBuffer.arrayOffset() + this.mBuffer.position());
            this.mBuffer = byteBufferAllocate;
        } else if (this.mBuffer.capacity() > 4096 && i < MIN_BLOB_SIZE) {
            ByteBuffer byteBufferAllocate2 = ByteBuffer.allocate(MIN_BLOB_SIZE);
            byteBufferAllocate2.put(this.mBuffer.array(), 0, this.mBuffer.arrayOffset() + this.mBuffer.position());
            this.mBuffer = byteBufferAllocate2;
        }
        read(this.mBuffer, i);
        this.mCRCBuf.clear();
        read(this.mCRCBuf, 4);
        this.mCRCBuf.position(0);
        int i2 = this.mCRCBuf.getInt();
        this.mChecksumTool.reset();
        this.mChecksumTool.update(this.mBuffer.array(), 0, this.mBuffer.position());
        if (i2 == ((int) this.mChecksumTool.getValue())) {
            byte[] bArr = this.mKey;
            if (bArr != null) {
                RC4Cryption.encrypt(bArr, this.mBuffer.array(), true, iPosition, i);
            }
            return this.mBuffer;
        }
        MyLog.w("CRC = " + ((int) this.mChecksumTool.getValue()) + " and " + i2);
        throw new IOException("Corrupted Blob bad CRC");
    }

    Blob read() throws IOException {
        int i = 0;
        try {
            ByteBuffer byteBufferReadOnePacket = readOnePacket();
            i = byteBufferReadOnePacket.position();
            byteBufferReadOnePacket.flip();
            byteBufferReadOnePacket.position(8);
            Blob ping = i == 8 ? new Ping() : Blob.from(byteBufferReadOnePacket.slice());
            MyLog.v("[Slim] Read {cmd=" + ping.getCmd() + ";chid=" + ping.getChannelId() + ";len=" + i + "}");
            return ping;
        } catch (IOException e) {
            if (i == 0) {
                i = this.mBuffer.position();
            }
            int i2 = i;
            if (i > 128) {
                i2 = 128;
            }
            MyLog.w("[Slim] read Blob [" + DebugUtils.bytes2Hex(this.mBuffer.array(), 0, i2) + "] Err:" + e.getMessage());
            throw e;
        }
    }

    void shutdown() {
        this.mDone = true;
    }

    void start() throws IOException {
        try {
            loop();
        } catch (IOException e) {
            if (!this.mDone) {
                throw e;
            }
        }
    }
}

File: legacy-runtime/src/main/java/com/xiaomi/slim/BlobReader.kt
package com.xiaomi.slim

import com.xiaomi.channel.commonutils.logger.MyLog
import com.xiaomi.channel.commonutils.misc.DebugUtils
import com.xiaomi.push.protobuf.ChannelMessage
import com.xiaomi.push.service.*
import java.io.BufferedInputStream
import java.io.EOFException
import java.io.IOException
import java.io.InputStream
import java.nio.ByteBuffer
import java.util.zip.Adler32

/*
 * Stock reference: com.xiaomi.xmsf 7.4.67-C (versionCode 70004067),
 * split-XiaomiServiceFrameworkCN-master.apk sha256 444e9f128591e04e38672bfe44a246ab3fa97ae68e95882839d8a7afe766df2b,
 * JADX path: com.xiaomi.xmsf/stock/split-XiaomiServiceFrameworkCN-master/sources/pa/d.java
 * Stock class name is obfuscated as pa.d; this file keeps the deobfuscated com.xiaomi.slim.BlobReader API.
 */
internal class BlobReader(
    inputStream: InputStream,
    private val mConnection: SlimConnection,
) {
    private val mInputStream: InputStream = BufferedInputStream(inputStream)
    private val mPacketParser = PacketParser()
    private var mBuffer: ByteBuffer = ByteBuffer.allocate(MIN_BLOB_SIZE)
    private val mCRCBuf: ByteBuffer = ByteBuffer.allocate(4)
    private val mChecksumTool = Adler32()
    private var mKey: ByteArray? = null
    @Volatile
    private var mDone: Boolean = false

    private fun loop() {
        mDone = false
        var valid = false
        val blob = read()
        if (Blob.CMD_CONN == blob.cmd) {
            val from = ChannelMessage.XMMsgConnResp.parseFrom(blob.payload)
            val observer = XMPushServiceProxy.get()?.runtimeObserver
            val hasChallenge = from.hasChallenge()
            val hasConfigMessage = from.hasPsc()
            val handshakePlan = observer?.planSlimHandshake(hasChallenge, hasConfigMessage) ?: PushSlimHandshakePlan(true, "slim_handshake_sent", false)
            MyLog.w("[slim] ${handshakePlan.eventAction}")
            valid = handshakePlan.valid
            if (handshakePlan.valid) {
                 // Challenge is now handled via notification or direct field if needed
                 // mConnection.challenge is usually set during bind or from CONN resp
            }
            if (handshakePlan.shouldEmitConfigBlob) {
                val psc = from.psc
                val blob2 = Blob().apply {
                    setCmd(Blob.CMD_SYNC, Blob.SUBCMD_CONF)
                    setPayload(psc.toByteArray(), null)
                }
                mConnection.notifyDataArrived(blob2)
            }
            MyLog.w("[Slim] CONN: host = ${from.host}")
            mConnection.notifyDataArrived(blob)
        }
        if (!valid) {
            MyLog.w("[Slim] Invalid CONN")
            throw IOException("Invalid Connection")
        }
        mKey = mConnection.key
        while (!mDone) {
            val blob3 = read()
            mConnection.setReadAlive()
            val observer = XMPushServiceProxy.get()?.runtimeObserver
            val payloadPlan = observer?.resolveSlimInboundPlan(blob3.channelId, blob3.cmd) ?: PushSlimInboundPlan(PushSlimInboundAction.None)
            
            payloadPlan.eventAction?.let { MyLog.w("[slim] $it") }
            
            when (payloadPlan.action) {
                PushSlimInboundAction.DeliverBlob -> {
                    mConnection.notifyDataArrived(blob3)
                }
                PushSlimInboundAction.ParseSecurePacket -> {
                    val packageName = blob3.packageName
                    try {
                        val key = mConnection.key
                        if (key != null) {
                            RC4Cryption.encrypt(key, blob3.payload, true, 0, blob3.payload.size)
                            mConnection.notifyDataArrived(mPacketParser.parse(blob3.payload, mConnection))
                        } else {
                            mConnection.notifyDataArrived(mPacketParser.parse(blob3.getDecryptedPayload(mConnection.challenge), mConnection))
                        }
                    } catch (e: Exception) {
                        MyLog.e("fail to decrypt blob from $packageName. chid=${blob3.channelId} $e")
                    }
                }
                PushSlimInboundAction.ParsePacket -> {
                    try {
                        mConnection.notifyDataArrived(mPacketParser.parse(blob3.payload, mConnection))
                    } catch (e: Exception) {
                        MyLog.e("fail to parse blob chid=${blob3.channelId} $e")
                    }
                }
                else -> {
                    if (payloadPlan.shouldLogUnknownType) {
                        MyLog.w("unknown blob type chid=${blob3.channelId}")
                    }
                }
            }
        }
    }

    @Throws(IOException::class)
    private fun read(
        byteBuffer: ByteBuffer,
        length: Int,
    ) {
        var remaining = length
        var position = byteBuffer.position()
        do {
            val read = mInputStream.read(byteBuffer.array(), position, remaining)
            if (read == -1) {
                throw EOFException()
            }
            remaining -= read
            position += read
        } while (remaining > 0)
        byteBuffer.position(position)
    }

    @Throws(IOException::class)
    private fun readOnePacket(): ByteBuffer {
        mBuffer.clear()
        read(mBuffer, 8)
        val s = mBuffer.getShort(0)
        val s2 = mBuffer.getShort(2)
        if (s != Blob.MAGIC || s2 != Blob.VERSION) {
            throw IOException("Malformed Input")
        }
        val i = mBuffer.getInt(4)
        val position = mBuffer.position()
        if (i > Blob.MAX_BLOB_SIZE) {
            throw IOException("Blob size too large")
        }
        if (i + 4 > mBuffer.remaining()) {
            val newBuffer = ByteBuffer.allocate(i + MIN_BLOB_SIZE)
            newBuffer.put(mBuffer.array(), 0, mBuffer.position())
            mBuffer = newBuffer
        } else if (mBuffer.capacity() > MAX_BLOB_THRSHOLD && i < MIN_BLOB_SIZE) {
            val newBuffer = ByteBuffer.allocate(MIN_BLOB_SIZE)
            newBuffer.put(mBuffer.array(), 0, mBuffer.position())
            mBuffer = newBuffer
        }
        read(mBuffer, i)
        mCRCBuf.clear()
        read(mCRCBuf, 4)
        mCRCBuf.position(0)
        val crcValue = mCRCBuf.int
        mChecksumTool.reset()
        mChecksumTool.update(mBuffer.array(), 0, mBuffer.position())
        if (crcValue == mChecksumTool.value.toInt()) {
            val key = mKey
            if (key != null) {
                RC4Cryption.encrypt(key, mBuffer.array(), true, position, i)
            }
            return mBuffer
        }
        MyLog.w("CRC = ${mChecksumTool.value.toInt()} and $crcValue")
        throw IOException("Corrupted Blob bad CRC")
    }

    @Throws(IOException::class)
    fun read(): Blob {
        var readLen = 0
        return try {
            val buffer = readOnePacket()
            readLen = buffer.position()
            buffer.flip()
            buffer.position(8)
            val blob = if (readLen == 8) Ping() else Blob.from(buffer.slice())
            MyLog.v("[Slim] Read {cmd=${blob.cmd};chid=${blob.channelId};len=$readLen}")
            blob
        } catch (e: IOException) {
            var len = if (readLen == 0) mBuffer.position() else readLen
            if (len > 128) len = 128
            MyLog.w("[Slim] read Blob [${DebugUtils.bytes2Hex(mBuffer.array(), 0, len)}] Err:${e.message}")
            throw e
        }
    }

    fun shutdown() {
        mDone = true
    }

    @Throws(IOException::class)
    fun start() {
        try {
            loop()
        } catch (e: IOException) {
            if (!mDone) throw e
        }
    }

    private companion object {
        private const val MAX_BLOB_THRSHOLD = 4096
        private const val MIN_BLOB_SIZE = 2048
    }
}
